typing/rec_check.ml

```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
         Dynamic
   in classify_expression Ident.empty
 
-let rec expression : Env.env -> Typedtree.expression -> Use.t =
-  fun env exp -> match exp.exp_desc with
+let remove_pat pat env =
+  Env.remove_list (pat_bound_idents pat) env
+
+let remove_patlist pats env =
+  List.fold_right remove_pat pats env
+
+(* Usage mode judgments.
+
+   There are two main groups of judgment functions:
```

Comment:
It took me a little while to work out what was going on here.  (It would have helped if I'd remembered to look at the clearer description in the PR when reading the code, but I'm glad I didn't, since it's important that the comments in the code are at least as clear as the PR.)  I think making the comment even more comprehensive could help.  You currently say

>   - Judgments of the form "Gamma |- ... : m"
>     compute the environment of a subterm from its mode,

but it would be a little bit clearer to say 

>   - Judgments of the form "G |- e : m"
>     compute the environment G given a subterm e and mode m,

and similarly for the second form of judgement.  (Note: it's clearer to consistently use `Gamma` or `G`, and to use the names of the various components whenever you mention them.

Aside: I wonder whether there's a name for the principality property that we want the type system to have?  Given `Γ ⊢ e : τ` we say that the system has "principal types" if there's always a most general `τ` given `Γ` and `e`, and we say that it has "principal typings" if there's always a most general `Γ` and `τ` given `e`.  Here we want there to be a most general `Γ` given `e` and `τ`.

Reply:
What about "principal contexts"?

I believe that this form of principality is a consequence of monotonicity (if G1 < G2 and `G2 |- ..` then `G1 |- ...`) and join preservation (if `G1 | - ...` and `G2 |- ...` then `G1 + G2 |- ...) -- note that the notation `m + m'` already assumes that two modes have a *unique* supremum. 

This system has very robust ordering properties: it is also monotonous on the `tau` side, using a weaker mode them will preserve the judgment, and compatibly with meet/min/`*` on that side (this is trivial as our order is linear), so as a consequence there is also a best/principal answer to the `G |- e : ?` question.

Reply:
I took your recommendations into account, and I also pushed a more comprehensive high-level documentation comment at the beginning of the file. The present comment was intended to be an implementation comment, explaining the type schemas of the functions extracted from term and binding judgments, not an explain-everything comment.